code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/integration/azure/AzureRESTClient.java

```diff
@@ -0,0 +1,431 @@
+            body.append(URLEncoder.encode(username, "UTF-8"));
+            body.append("&password=");
+            body.append(URLEncoder.encode(split[1], "UTF-8")); 
+            
+            HttpResponse response = Request.Post(AzureConstants.AUTHORITY)
+                    .addHeader(AzureConstants.ACCEPT, AzureConstants.APPLICATION_FORM_URL_ENCODED)
+                    .bodyString(body.toString(), ContentType.APPLICATION_FORM_URLENCODED).execute().returnResponse();
+            int statusCode = response.getStatusLine().getStatusCode();
+            
+            if(statusCode == 400) {
+                if(isUnAuthorizedError(response)) {
+                    throw new ClientVisibleException(ResponseCodes.UNAUTHORIZED);
+                }
+            }
+
+            if(statusCode >= 300) {
+                noAzure(statusCode);
+            }
+
+            Map<String, Object> jsonData = jsonMapper.readValue(response.getEntity().getContent());
+            accessToken = ObjectUtils.toString(jsonData.get("access_token"));
+            refreshToken = ObjectUtils.toString(jsonData.get("refresh_token"));
+            
+            ApiContext.getContext().getApiRequest().setAttribute(AzureConstants.AZURE_ACCESS_TOKEN, accessToken);
+            ApiContext.getContext().getApiRequest().setAttribute(AzureConstants.AZURE_REFRESH_TOKEN, refreshToken);
+
+        } catch(ClientVisibleException ex){
+            throw ex;
+        } catch(Exception ex) {
+            throw new RuntimeException(ex);
+        }
+
+        return accessToken;
+
+    }
+    
+    
+    public void refreshAccessToken() {
+        if (!isConfigured()) {
+            throw new ClientVisibleException(ResponseCodes.INTERNAL_SERVER_ERROR, AzureConstants.CONFIG, "Azure Client and Tenant Id not configured", null);
+        }
+        
+        logger.debug("Azure accessToken expired, refreshing the token");
+        
+        String azureAccessToken = (String) ApiContext.getContext().getApiRequest().getAttribute(AzureConstants.AZURE_ACCESS_TOKEN);
+        if (StringUtils.isEmpty(azureAccessToken)) {
+            noAccessToken();
+        }
+        
+        String azureRefreshToken = (String) ApiContext.getContext().getApiRequest().getAttribute(AzureConstants.AZURE_REFRESH_TOKEN);
+        if (StringUtils.isEmpty(azureRefreshToken)) {
+            noRefreshToken();
+        }
+        try{
+            //post to the microsoft azure authority
+            HttpResponse response = Request.Post(AzureConstants.AUTHORITY)
+                    .addHeader(AzureConstants.ACCEPT, AzureConstants.APPLICATION_FORM_URL_ENCODED)
+                    .bodyString("grant_type=refresh_token&refresh_token="+azureRefreshToken, ContentType.APPLICATION_FORM_URLENCODED)
+                    .execute().returnResponse();
+            int statusCode = response.getStatusLine().getStatusCode();
+            if(statusCode >= 300) {
+                noAzure(statusCode);
```

Comment:
looks like clientVisibleException thrown by noAzure, will be swallowed and transformed to runtime on line 167